src/validations/rules/ssp.sch

```diff
@@ -2209,38 +2209,61 @@
             <sch:assert
                 diagnostics="ipv4-has-content-diagnostic"
                 doc:guide-reference="Guide to OSCAL-based FedRAMP System Security Plans §6.5.1 and §6.5.2"
                 doc:template-reference="System Security Plan Template §15 Attachment 13"
                 id="ipv4-has-content"
                 role="error"
-                test="o:prop[@value ne '0.0.0.0' and normalize-space(@value) ne '' and @name eq 'ipv4-address']">
-                <xsl:value-of select="o:prop[@name='asset-id']/@value"/> does not have an appropriate IPv4 value.</sch:assert>
+                test="if(o:prop[@name eq 'ipv4-address'])
+                then(o:prop[matches(@value, '(^[0-9][0-9]?[0-9]?\.[0-9][0-9]?[0-9]?\.[0-9][0-9]?[0-9]?\.[0-9][0-9]?[0-9]?$)')])
+                else(true())"><xsl:value-of select="o:prop[@name='asset-id']/@value"/> must have an appropriate IPv4 value.</sch:assert>
+            <sch:assert
+                diagnostics="ipv4-has-non-placeholder-diagnostic"
+                doc:guide-reference="Guide to OSCAL-based FedRAMP System Security Plans §6.5.1 and §6.5.2"
+                doc:template-reference="System Security Plan Template §15 Attachment 13"
+                id="ipv4-has-non-placeholder"
+                role="error"
+                test="if(o:prop[@name eq 'ipv4-address'])
+                then(o:prop[matches(@value, '0.0.0.0')])
+                else(false())"><xsl:value-of select="o:prop[@name='asset-id']/@value"/> must have an appropriate IPv4 value.</sch:assert>
             <sch:assert
                 diagnostics="ipv6-has-content-diagnostic"
                 doc:guide-reference="Guide to OSCAL-based FedRAMP System Security Plans §6.5.1 and §6.5.2"
                 doc:template-reference="System Security Plan Template §15 Attachment 13"
                 id="ipv6-has-content"
                 role="error"
-                test="o:prop[@value ne '0000:0000:0000:0000' and normalize-space(@value) ne '' and @name eq 'ipv6-address']">
-                <xsl:value-of select="o:prop[@name='asset-id']/@value"/> does not have an appropriate IPv6 value.</sch:assert>
+                test="if(o:prop[@name eq 'ipv6-address'])
+                then(o:prop[matches(@value, '([A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}')])
+                else(true())"><xsl:value-of select="o:prop[@name='asset-id']/@value"/> must have an appropriate IPv6 value.</sch:assert>
+            <sch:assert
+                diagnostics="ipv6-has-non-placeholder-diagnostic"
+                doc:guide-reference="Guide to OSCAL-based FedRAMP System Security Plans §6.5.1 and §6.5.2"
+                doc:template-reference="System Security Plan Template §15 Attachment 13"
+                id="ipv6-has-non-placeholder"
+                role="error"
+                test="if(o:prop[@name eq 'ipv6-address'])
+                then(o:prop[matches(@value, '0000:0000:0000:0000:0000:0000:0000:0000')])
```

Comment:
`0000:0000:0000:0000:0000:0000:0000:0000` does not conform to [RFC 5942](https://datatracker.ietf.org/doc/html/rfc5952) and thus will be unlikely to be encountered. There are other representations of a "zero" IPv6 address.

As previously mentioned,
> For IPv6, see [RFC 4291 §2](https://datatracker.ietf.org/doc/html/rfc4291#section-2) for addresses such as `0:0:0:0:0:0:0:0` and `::` (and probably `::0`).

Reply:
Unlikely but not non-zero.  Thus having it causes no real impact and might identify errors where the regex would miss.  The suggested regexes, that are not working within the XPath, find the 0000, etc string as valid.  While it might be a valid IP it seems likely to be worthless for the ATO.

If there are other zero representations of ipv6 address please point me to them.

Reply:
IPv6 addresses in the wild will have been "compressed" and thus `0000:0000:0000:0000:0000:0000:0000:0000` would appear as `::0` or perhaps even `::`.

To add precision, there are multiple IPv6 address variants which should be considered suspect - unacceptable - in an SSP inventory. One of these is what had been referred to as a "_zeros placeholder_" address in the hastily-specified backlog issue. The intent of #358 was to validate M-21-07 compliance as well as to reject bogus network addresses.

For IPv6, the following (RFC 5942) forms minimally should be rejected - not just _zeros placeholder_:

- `::` (this is the so-called _zeros placeholder_ address)
- `::0` (a variant of _zeros placeholder_)
- `0000:0000:0000:0000:0000:0000:0000:0000` **with zero string cardinality from 1 to 4** in any of the segments (aka _zeros placeholder_)
- `::1` (loopback)
- `::ffff/96`
- `fc00::/7` (probably)
- `fe80::/10` (probably)
- `2001:0000::/32` Teredo
- `2001:0002::/48`
- `2001:0010::/28`
- `2002::/16`
- `2001:db8::/32` (documentation)
- `2002::/16`
- `fc00::/7`
- `fe80::/` (link local)
and likely more.

A safe bet would be to just reject the RFC 5942 form - `::` and perhaps the explicit `::0` variants of the _zeros placeholder_ address.

Reply:
I think this is now moot since the change for storing the regex in a variable fixed the issue.

Reply:
The regex at line 2242 in the current branch recognizes correctly-formed IPv6 addresses, including `0000:0000:0000:0000:0000:0000:0000:0000` and `::`.

The regex at line 2259 (`0000:0000:0000:0000:0000:0000:0000:0000`) in the current branch's ssp.sch assertion `ipv6-has-non-placeholder` will not detect all zeros placeholder entries. In particular, it would miss any variant which removed an arbitrary number of leading zeros in an arbitrary number of segments. The actual all-zeros IPv6 address in common compressed form is simply `::`.